projects/RabbitMQ.Client/client/impl/ConsumerWorkService.cs

```diff
@@ -105,10 +124,54 @@ async Task Loop()
                 }
             }
 
+            async Task LoopWithConcurrency(CancellationToken cancellationToken)
+            {
+                while (_tokenSource.IsCancellationRequested == false)
+                {
+                    try
+                    {
+                        await _syncSource.Task.ConfigureAwait(false);
+                        _syncSource = new TaskCompletionSource<bool>(TaskCreationOptions.RunContinuationsAsynchronously);
+                    }
+                    catch (TaskCanceledException)
+                    {
+                        // Swallowing the task cancellation exception for the semaphore in case we are stopping.
+                    }
+
+                    while (_actions.TryDequeue(out Action action))
+                    {
+                        // Do a quick synchronous check before we resort to async/await with the state-machine overhead.
+                        if(!_limiter.Wait(0))
+                        {
+                            await _limiter.WaitAsync(cancellationToken).ConfigureAwait(false);
+                        }
+
+                        _ = OffloadToWorkerThreadPool(action, _limiter);
+                    }
+                }
```

Comment:
Will these two loops spin continuously (causing CPU load) since there is no Delay/Sleep between iterations when the queue is empty? Not sure of the mechanisms in the ConcurrentQueue implementation.

Reply:
Would it make sense to have the `Enqueue` method trigger the loop when needed via. AutoResetEvent instead?

Reply:
The behavior of the `LoopWithConcurrency` is the same as `Loop` which is the existing behavior. The loop sits idle until something is in the queue. Once there is something in the queue it consumes as long as there is something to consume and then is idle again until there is something enqueued. It is correct that continues scheduling of work will "capture" the worker in the inner while. At this stage it is by design so I feel any discussions about changing that behavior should probably be separate from this PR?

Reply:
Yeah, was just wondering because the AsyncConsumerWorkService has a Channel as a work item buffer where it can properly await until a message arrives in the channel, where the normal ConsumerWorkService is using the ConcurrentQueue instead. Just wondering if the rapid iteration in TryDequeue (which will return false if the queue is empty, causing a spinning loop until there is an item in there) will cause unneeded CPU usage. Couldn't that just as well be replaced with the Channel implementation as well, making the WorkServices pretty much identical apart from that one execute a synchronous action and the other one a Task based action?

Reply:
It doesn't case a spinning loop if there is nothing in the queue because the inner while is exited and then 

```
await _syncSource.Task.ConfigureAwait(false);
```

kick in which is not reset because there was nothing added to the queue. Therefore it just sits there. In case there was something added to the queue it falls through, set the new TCS and then dequeues until done.

> Couldn't that just as well be replaced with the Channel implementation as well, making the WorkServices pretty much identical apart from that one execute a synchronous action and the other one a Task based action?

Sure that would be possible and always was even when it was decided to diverge the async service to use channel. I'm fine to do this at a later stage, still I found it not relevant right now for the scope of this PR. If you have good arguments why those two changes should be bundled I'm happy to listen though

Reply:
> It doesn't case a spinning loop if there is nothing in the queue because the inner while is exited and then
> 
> ```
> await _syncSource.Task.ConfigureAwait(false);
> ```

Guess I'm not awake yet... I completely missed that await :) The other changes can wait from my point of view, this should work just fine.